### ISSUE TITLE
DS Prefill :: Investigate row-parallel vs column-parallel LM head + perf comparison

### ISSUE THREAD

**Issue by mbezuljTT** [tenstorrent/tt-metal#43131]
## Context

Current LM head uses column-parallel: all-gather input across TP to get full emb_dim, then matmul with vocab-sharded weight `[emb, vocab/tp]`. Output is TP-sharded on vocab, requiring concat to reconstruct full vocab for sampling.

## Alternative Approach (Row-Parallel)

Keep input TP-sharded `[TILE, emb/tp]`, use emb-sharded weight `[emb/tp, vocab]`, matmul gives partial sum `[TILE, vocab]`, then all-reduce across TP. Output is replicated with full vocab — no concat needed for sampling.

## Action Items

- [ ] Implement row-parallel variant
- [ ] Perf benchmark both approaches (all-gather input vs all-reduce output)
- [ ] Evaluate memory and communication tradeoffs

## Tradeoffs

| Approach | Communication | Output |
|----------|--------------|--------|
| Column-parallel (current) | All-gather input (emb_dim) | TP-sharded on vocab, needs concat |
| Row-parallel | All-reduce output (vocab) | Replicated, ready for sampling |

The all-reduce output may be larger (full vocab) vs all-gather input (full emb_dim), but eliminates the need for host-side concatenation before sampling.

**Comment by nbabinTT** [comment]
I implemented row-parallel LM head, and measured performance of both approaches. Implementation of the row-parallel approach is available at [nbabin/lm_head_row_parallel](https://github.com/tenstorrent/tt-metal/tree/nbabin/lm_head_row_parallel) branch. All measurements were done on 25k ISL on a single Galaxy.

I measured the perf of both approaches and got the following numbers:
**1. Column-parallel approach**

<img width="1208" height="339" alt="Image" src="https://github.com/user-attachments/assets/c45db232-7ced-4443-a0fe-9b5273a88919" />


**2. Row-parallel approach**

<img width="1208" height="371" alt="Image" src="https://github.com/user-attachments/assets/190a5f5f-9987-4c08-8a35-de4ad8b7b577" />

From the perf-report we see that Matmul ops in both cases are DRAM bound and the kernel duration is very similar. The difference between these two approaches are All Gather and All Reduce ops (in the perf report, All Reduce consists of Reduce Scatter and All Gather ops) that are applied to input and output tensor, respectively. As per perf report, duration of All Gather in column-parallel case is 38us compared to the 184us (RS) + 165us (AG) in case of the row-parallel variant.

Amount of data transferred in case of **column-parallel:**
1. All Gather
```
input_0  = [1, 1, 32, 1792]    bf16
output_0 = [1, 1, 32, 7168]    bf16
topology = Linear

input_size  = 1 × 1 × 32 × 1792  × 2 = 114,688 B
output_size = 1 × 1 × 32 × 7168  × 2 = 458,752 B

nodes = 4
tensor_size = max(input_size, output_size) = 458,752 B
all_gather_bandwidth_ring = (nodes - 1) x tensor_size / nodes
data_to_move_ring = 458,752 × 3/4 = 344,064 B
data_to_move_linear = 344,064 × 2 = 688,128 B

Total bytes moved in case of line topology: 688,128 B ≈ 688 KB per device
```

Amount of data transferred in case of **row-parallel:**
1. Reduce Scatter
```
input_0  = [1, 1, 32, 129280]  bf16    ← partial sum, full vocab
output_0 = [1, 1, 32,  32320]  bf16    ← scattered slice, vocab/tp
topology = Linear

input_size  = 1 × 1 × 32 × 129280 × 2 = 8,273,920 B
output_size = 1 × 1 × 32 ×  32320 × 2 = 2,068,480 B

nodes = 4
tensor_size = max(input_size, output_size) = 8,273,920 B
reduce_scatter_bandwidth_ring = (nodes - 1) x tensor_size / nodes
data_to_move_ring = 8,273,920 × 3/4 = 6,205,440 B
data_to_move_linear = 6,205,440 × 2 = 12,410,880 B

Total bytes moved in case of line topology: 12,410,880 B ≈ 12.4 MB per device
```

2. All Gather
```
input_0  = [1, 1, 32,  32320]  bf16
output_0 = [1, 1, 32, 129280]  bf16
topology = Linear

input_size  = 1 × 1 × 32 ×  32320 × 2 = 2,068,480 B
output_size = 1 × 1 × 32 × 129280 × 2 = 8,273,920 B

nodes = 4
tensor_size = max(input_size, output_size) = 8,273,920 B
all_gather_bandwidth_ring = (nodes - 1) x tensor_size / nodes
data_to_move_ring = 8,273,920 × 3/4 = 6,205,440 B
data_to_move_linear = 6,205,440 × 2 = 12,410,880 B

Total bytes moved in case of line topology: 12,410,880 B ≈ 12.4 MB per device
```

Based on these calculations, data moved between devices in case of column parallel is 36 times smaller compared to the row-parallel case leading to more communication between devices.

I also measured cost of transferring logits from device to host. In column-parallel case, I measured 360ms, and in case of row-parallel approach, I measured 370ms. The problem lies in that mash composer reads tensors from all devices and then does concatenation/slicing on the host. So effectively, instead of reading just one column of devices on Galaxy (8 devices), all 32 devices are read from in row-parallel case (column-parallel case requires reading from all 32 devices). I will check if there are any alternatives to it.

**Comment by nbabinTT** [comment]
I managed to measure time of transferring data from 8 devices to host in case of the row-parallel case. I've done by calling to_torch 8 times targeting chips from the first column. In that case, I got ~80ms compared to the 360ms I got with mesh composer which is 4.5 times faster.

| | Column-parallel | Row-parallel |
| :------- | :-------: | :------: |
| CCL ops | 38 us | 184 us + 165 us = 349 us |
| Device to host | 365 ms | 80 ms |
| Total | 365.04 ms | 80.35 ms |

**Comment by nbabinTT** [comment]
Managed to optimize device to host transfers by reducing the number of chips we read from. In case of column-parallel approach, we now read only the row of devices that have the token of interest - tensors from those 4 devices are transferred to host where they are concatenated. In row-parallel case, we need to read from only one device in row that contains the token of interest. We read from only one device as data is replicated on all devices across columns.

| | Column-parallel | Row-parallel |
| :------- | :-------: | :------: |
| CCL ops | 38 us | 184 us + 165 us = 349 us |
| Optimized device to host | 13.60 ms | 6.67 ms |
| Total | 13.64 ms | 7.02 ms |

Row-parallel approach still seems a way to go as end to end time benefits from reduced host to device transfer time.
